### PR TITLE
dev/core#2644 Add Scheduled Communications extension

### PR DESCRIPTION
**Pull request by colemanw**
Overview
----------------------------------------
Allows Scheduled Reminders to be created from SearchKit with an optional extension.

Before
-------
Only a handful of Scheduled Reminder entities available (Contact, Participant, Member, Activity), with narrowly-restricted criteria.

After
--------
When enabled, the "Scheduled Communications" extension allows any SearchKit search with a date column and a contact id column can be used for scheduled reminders.

Comments
---------
See https://lab.civicrm.org/dev/core/-/issues/2644

### CODE REVIEW

**Comment by colemanw** [comment]
@eileenmcnaughton this is ready for review
retest this please

**Comment by colemanw** [comment]
retest this please

**Comment by eileenmcnaughton** [comment]
Flushing system caches                                                       
  ~/buildkit/build/build-0/web/sites/all/modules/civicrm/ext/search_kit ~/bui  
  ldkit                                                                        
  TAP version 13                                                               
  PHPUnit 8.5.27 #StandWithUkraine                            
                                                                               
                                                                               
  Warning: Undefined array key "Civi\Core\Container" in /home/homer/buildkit/  
  build/build-0/web/sites/all/modules/civicrm/Civi/Core/Container.php on line  
   702                                                                         
                                                                               
  Warning: Trying to access array offset on value of type null in /home/homer  
  /buildkit/build/build-0/web/sites/all/modules/civicrm/Civi/Core/Container.p  
  hp on line 702                                                               
                                                                               
  Warning: Trying to access array offset on value of type null in /home/homer  
  /buildkit/build/build-0/web/sites/all/modules/civicrm/Civi/Core/Container.p  
  hp on line 702                                                               
                                                                               
  Fatal error: Uncaught Error: Call to a member function getPath() on null in  
   /home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/Civi/Test  
  /EventChecker.php:94                                                         
  Stack trace:                                                                 
  #0 /home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/Civi/Te  
  st/EventChecker.php(36): Civi\Test\EventChecker->findAll()                   
  #1 /home/homer/buildkit/build/build-0/web/sites/all/modules/civicrm/Civi/Te  
  st/CiviTestListenerPHPUnit7.php(64): Civi\Test\EventChecker->start(Object(C  
  ivi\Search\ActionMappingTest))                                               
  #2 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/Framew  
  ork/TestResult.php(362): Civi\Test\CiviTestListenerPHPUnit7->startTest(Obje  
  ct(Civi\Search\ActionMappingTest))                                           
  #3 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/Framew  
  ork/TestSuite.php(453): PHPUnit\Framework\TestResult->startTest(Object(Civi  
  \Search\ActionMappingTest))                                                  
  #4 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/Framew  
  ork/TestSuite.php(475): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Fra  
  mework\TestResult))                                                          
  #5 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/TextUI  
  /TestRunner.php(462): PHPUnit\Framework\TestSuite->run(Object(PHPUnit\Frame  
  work\TestResult))                                                            
  #6 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/TextUI  
  /Command.php(129): PHPUnit\TextUI\TestRunner->doRun(Object(PHPUnit\Framewor  
  k\TestSuite), Array, Array, true)                                            
  #7 phar:///home/homer/buildkit/extern/phpunit8/phpunit8.phar/phpunit/TextUI  
  /Command.php(101): PHPUnit\TextUI\Command->run(Array, true)                  
  #8 /home/homer/buildkit/extern/phpunit8/phpunit8.phar(1721): PHPUnit\TextUI  
  \Command::main()                                                             
  #9 {main}                                                                    
    thrown in /home/homer/buildkit/build/build-0/web/sites/all/modules/civicr  
  m/Civi/Test/EventChecker.php on line 94

**Comment by colemanw** [comment]
retest this please

**Comment by colemanw** [comment]
@eileenmcnaughton I moved the test to a different directory and now it's passing 🤷🏻

File: Civi/ActionSchedule/RecipientBuilder.php [hunk 1]
```diff
@@ -402,7 +402,7 @@ protected function prepareStartDateClauses() {
       $date = $operator . "(!casDateField, INTERVAL {$actionSchedule->start_action_offset} {$actionSchedule->start_action_unit})";
       $startDateClauses[] = "'!casNow' >= {$date}";
       // This is weird. Waddupwidat?
-      if ($this->mapping->getEntityTable() == 'civicrm_participant') {
+      if ($this->getMappingTable() == 'civicrm_participant') {
```
Comment (eileenmcnaughton) [inline Civi/ActionSchedule/RecipientBuilder.php:405]: Non-blocking but === would be better

**Comment by eileenmcnaughton** [comment]
@colemanw I gave this a spin & was able to configure & test a Search-kit based scheduled reminders which went OK, and did resolve tokens.

Given the scope of this PR is mostly internal to this work & there is a lot of code I'm OK with it being merged as is & building on it - however, I think we need to ensure there are some safeguards in place in the SAME release this goes out in. The minimum version I can think of is 

1) add new permission - administer search kit scheduled reminders
2) option to configure only available in search kit UI for users with this permission
3) saved searches with admin ui can't be modified by someone without the permission
4) saved search field frozen for users without the permission (it feels risky to me being able to change that once created anyway - I kinda lean to making people delete & recreate)
image

**Comment by JoeMurray** [comment]
Just checking that the new perm is validated for api calls as well as browser based actions.

**Comment by eileenmcnaughton** [comment]
@colemanw as my non-trusted user (has lots of administer but not the magic tier 1 or the new perm) I still see the search (which is fine) - however, I do not expect this user to be able to delete the communication - which they can

image

image

image

**Comment by eileenmcnaughton** [comment]
On the non-admin user delete the database still has the action_schedule entry (the second one)

image

**Comment by eileenmcnaughton** [comment]
However the screen fails to load with it deleted 

image

**Comment by eileenmcnaughton** [comment]
Also, in my testing the orphan scheduled reminder did not go out, but it did block the non-saved search one too & this could be bad. I think this is a high risk area & it would be good to cover the specific scenario of an orphan sk communiction/reminder in a unit test

**Comment by eileenmcnaughton** [comment]
I think it might be worth adding a check on the `expires_date` field to the disabled / deleted search display test. There doesn't seem to be an is_active in civicrm_saved_search

**Comment by eileenmcnaughton** [comment]
Hmm this is tricksy - I'm getting an invalid field 'is_deceased'


My search

```
[
  [
    "SavedSearch",
    "save",
    {
      "records": [
        {
          "name": "big_money",
          "label": "big money",
          "form_values": null,
          "mapping_id": null,
          "search_custom_id": null,
          "api_entity": "Contribution",
          "api_params": {
            "version": 4,
            "select": [
              "contact_id.sort_name",
              "total_amount",
              "financial_type_id:label",
              "contribution_status_id:label",
              "id",
              "receive_date"
            ],
            "orderBy": [],
            "where": [
              [
                "total_amount",
                ">",
                "1000"
              ]
            ],
            "groupBy": [],
            "join": [],
            "having": []
          },
          "expires_date": null,
          "description": null
        }
      ],
      "match": [
        "name"
      ]
    }
  ]
]
```

<tr><th align='left' bgcolor='#f57900' colspan="5"><span style='background-color: #cc0000; color: #fce94f; font-size: x-large;'>( ! )</span> CRM_Core_Exception: Invalid field 'is_deceased' in /srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php on line <i>372</i></th></tr>
<tr><th align='left' bgcolor='#e9b96e' colspan='5'>Call Stack</th></tr>
<tr><th align='center' bgcolor='#eeeeec'>#</th><th align='left' bgcolor='#eeeeec'>Time</th><th align='left' bgcolor='#eeeeec'>Memory</th><th align='left' bgcolor='#eeeeec'>Function</th><th align='left' bgcolor='#eeeeec'>Location</th></tr>
<tr><td bgcolor='#eeeeec' align='center'>1</td><td bgcolor='#eeeeec' align='center'>0.0046</td><td bgcolor='#eeeeec' align='right'>381952</td><td bgcolor='#eeeeec'>{main}(  )</td><td title='/srv/civi-sites/dmaster/web/index.php' bgcolor='#eeeeec'>.../index.php<b>:</b>0</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>2</td><td bgcolor='#eeeeec' align='center'>0.1092</td><td bgcolor='#eeeeec' align='right'>1501712</td><td bgcolor='#eeeeec'>menu_execute_active_handler( <span>$path = </span>???, <span>$deliver = </span>??? )</td><td title='/srv/civi-sites/dmaster/web/index.php' bgcolor='#eeeeec'>.../index.php<b>:</b>21</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>3</td><td bgcolor='#eeeeec' align='center'>0.1094</td><td bgcolor='#eeeeec' align='right'>1502256</td><td bgcolor='#eeeeec'>civicrm_invoke( <span>&#39;ajax&#39;</span>, <span>&#39;rest&#39;</span> )</td><td title='/srv/civi-sites/dmaster/web/includes/menu.inc' bgcolor='#eeeeec'>.../menu.inc<b>:</b>527</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>4</td><td bgcolor='#eeeeec' align='center'>0.2747</td><td bgcolor='#eeeeec' align='right'>3107184</td><td bgcolor='#eeeeec'>CRM_Core_Invoke::invoke( <span>$args = </span><span>[0 =&gt; &#39;civicrm&#39;, 1 =&gt; &#39;ajax&#39;, 2 =&gt; &#39;rest&#39;]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/drupal/civicrm.module' bgcolor='#eeeeec'>.../civicrm.module<b>:</b>472</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>5</td><td bgcolor='#eeeeec' align='center'>0.2747</td><td bgcolor='#eeeeec' align='right'>3107184</td><td bgcolor='#eeeeec'>CRM_Core_Invoke::_invoke( <span>$args = </span><span>[0 =&gt; &#39;civicrm&#39;, 1 =&gt; &#39;ajax&#39;, 2 =&gt; &#39;rest&#39;]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/Invoke.php' bgcolor='#eeeeec'>.../Invoke.php<b>:</b>36</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>6</td><td bgcolor='#eeeeec' align='center'>0.2878</td><td bgcolor='#eeeeec' align='right'>3156832</td><td bgcolor='#eeeeec'>CRM_Core_Invoke::runItem( <span>$item = </span><span>[&#39;id&#39; =&gt; &#39;147&#39;, &#39;domain_id&#39; =&gt; &#39;1&#39;, &#39;path&#39; =&gt; &#39;civicrm/ajax/rest&#39;, &#39;access_callback&#39; =&gt; [0 =&gt; &#39;CRM_Core_Permission&#39;, 1 =&gt; &#39;checkMenu&#39;], &#39;access_arguments&#39; =&gt; [0 =&gt; [...], 1 =&gt; &#39;or&#39;], &#39;page_callback&#39; =&gt; [0 =&gt; &#39;CRM_Utils_REST&#39;, 1 =&gt; &#39;ajax&#39;], &#39;breadcrumb&#39; =&gt; [0 =&gt; [...]], &#39;is_active&#39; =&gt; &#39;1&#39;, &#39;is_public&#39; =&gt; &#39;0&#39;, &#39;is_exposed&#39; =&gt; &#39;1&#39;, &#39;is_ssl&#39; =&gt; &#39;0&#39;, &#39;weight&#39; =&gt; &#39;1&#39;, &#39;type&#39; =&gt; &#39;1&#39;, &#39;page_type&#39; =&gt; &#39;0&#39;, &#39;skipBreadcrumb&#39; =&gt; &#39;0&#39;, &#39;page_arguments&#39; =&gt; FALSE]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/Invoke.php' bgcolor='#eeeeec'>.../Invoke.php<b>:</b>69</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>7</td><td bgcolor='#eeeeec' align='center'>0.3055</td><td bgcolor='#eeeeec' align='right'>3523056</td><td bgcolor='#eeeeec'>CRM_Utils_REST::ajax(  )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/Invoke.php' bgcolor='#eeeeec'>.../Invoke.php<b>:</b>285</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>8</td><td bgcolor='#eeeeec' align='center'>0.3057</td><td bgcolor='#eeeeec' align='right'>3523840</td><td bgcolor='#eeeeec'>CRM_Utils_REST::process( <span>$args = </span><span>[0 =&gt; &#39;civicrm&#39;, 1 =&gt; &#39;job&#39;, 2 =&gt; &#39;send_reminder&#39;]</span>, <span>$params = </span><span>[]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Utils/REST.php' bgcolor='#eeeeec'>.../REST.php<b>:</b>533</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>9</td><td bgcolor='#eeeeec' align='center'>0.3057</td><td bgcolor='#eeeeec' align='right'>3524256</td><td bgcolor='#eeeeec'>civicrm_api( <span>$entity = </span><span>&#39;job&#39;</span>, <span>$action = </span><span>&#39;send_reminder&#39;</span>, <span>$params = </span><span>[&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Utils/REST.php' bgcolor='#eeeeec'>.../REST.php<b>:</b>288</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>10</td><td bgcolor='#eeeeec' align='center'>0.3115</td><td bgcolor='#eeeeec' align='right'>3552288</td><td bgcolor='#eeeeec'>Civi\API\Kernel->runSafe( <span>$entity = </span><span>&#39;job&#39;</span>, <span>$action = </span><span>&#39;send_reminder&#39;</span>, <span>$params = </span><span>[&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/api/api.php' bgcolor='#eeeeec'>.../api.php<b>:</b>28</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>11</td><td bgcolor='#eeeeec' align='center'>0.3118</td><td bgcolor='#eeeeec' align='right'>3552752</td><td bgcolor='#eeeeec'>Civi\API\Kernel->runRequest( <span>$apiRequest = </span><span>[&#39;id&#39; =&gt; 1, &#39;version&#39; =&gt; 3, &#39;params&#39; =&gt; [&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3], &#39;fields&#39; =&gt; NULL, &#39;entity&#39; =&gt; &#39;Job&#39;, &#39;action&#39; =&gt; &#39;send_reminder&#39;]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/API/Kernel.php' bgcolor='#eeeeec'>.../Kernel.php<b>:</b>79</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>12</td><td bgcolor='#eeeeec' align='center'>0.3243</td><td bgcolor='#eeeeec' align='right'>3678344</td><td bgcolor='#eeeeec'>Civi\API\Provider\MagicFunctionProvider->invoke( <span>$apiRequest = </span><span>[&#39;id&#39; =&gt; 1, &#39;version&#39; =&gt; 3, &#39;params&#39; =&gt; [&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3], &#39;fields&#39; =&gt; [], &#39;entity&#39; =&gt; &#39;Job&#39;, &#39;action&#39; =&gt; &#39;send_reminder&#39;, &#39;function&#39; =&gt; &#39;civicrm_api3_job_send_reminder&#39;, &#39;is_generic&#39; =&gt; FALSE]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/API/Kernel.php' bgcolor='#eeeeec'>.../Kernel.php<b>:</b>156</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>13</td><td bgcolor='#eeeeec' align='center'>0.3243</td><td bgcolor='#eeeeec' align='right'>3678344</td><td bgcolor='#eeeeec'>civicrm_api3_job_send_reminder( <span>$params = </span><span>[&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/API/Provider/MagicFunctionProvider.php' bgcolor='#eeeeec'>.../MagicFunctionProvider.php<b>:</b>89</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>14</td><td bgcolor='#eeeeec' align='center'>0.3274</td><td bgcolor='#eeeeec' align='right'>3726048</td><td bgcolor='#eeeeec'>CRM_Core_BAO_ActionSchedule::processQueue( <span>$now = </span><span>NULL</span>, <span>$params = </span><span>[&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3, &#39;rowCount&#39; =&gt; 0]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/api/v3/Job.php' bgcolor='#eeeeec'>.../Job.php<b>:</b>204</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>15</td><td bgcolor='#eeeeec' align='center'>26.0499</td><td bgcolor='#eeeeec' align='right'>5321088</td><td bgcolor='#eeeeec'>CRM_Core_BAO_ActionSchedule::buildRecipientContacts( <span>$mappingID = </span><span>&#39;saved_search&#39;</span>, <span>$now = </span><span>&#39;20230830052037&#39;</span>, <span>$params = </span><span>[&#39;check_permissions&#39; =&gt; TRUE, &#39;version&#39; =&gt; 3, &#39;rowCount&#39; =&gt; 0]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/BAO/ActionSchedule.php' bgcolor='#eeeeec'>.../ActionSchedule.php<b>:</b>467</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>16</td><td bgcolor='#eeeeec' align='center'>27.9993</td><td bgcolor='#eeeeec' align='right'>5346048</td><td bgcolor='#eeeeec'>Civi\ActionSchedule\RecipientBuilder->build(  )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/CRM/Core/BAO/ActionSchedule.php' bgcolor='#eeeeec'>.../ActionSchedule.php<b>:</b>450</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>17</td><td bgcolor='#eeeeec' align='center'>30.3750</td><td bgcolor='#eeeeec' align='right'>5346048</td><td bgcolor='#eeeeec'>Civi\ActionSchedule\RecipientBuilder->buildRelFirstPass(  )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/ActionSchedule/RecipientBuilder.php' bgcolor='#eeeeec'>.../RecipientBuilder.php<b>:</b>139</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>18</td><td bgcolor='#eeeeec' align='center'>31.4242</td><td bgcolor='#eeeeec' align='right'>5346048</td><td bgcolor='#eeeeec'>Civi\ActionSchedule\RecipientBuilder->prepareQuery( <span>$phase = </span><span>&#39;rel-first&#39;</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/ActionSchedule/RecipientBuilder.php' bgcolor='#eeeeec'>.../RecipientBuilder.php<b>:</b>162</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>19</td><td bgcolor='#eeeeec' align='center'>31.5162</td><td bgcolor='#eeeeec' align='right'>5401704</td><td bgcolor='#eeeeec'>Civi\Search\ActionMapping->createQuery( <span>$schedule = </span><span>class CRM_Core_DAO_ActionSchedule { public $id = &#39;3&#39;; public $name = &#39;new_one&#39;; public $title = &#39;new one&#39;; public $recipient = NULL; public $limit_to = NULL; public $entity_value = &#39;23&#39;; public $entity_status = &#39;id&#39;; public $start_action_offset = &#39;1&#39;; public $start_action_unit = &#39;day&#39;; public $start_action_condition = &#39;after&#39;; public $start_action_date = &#39;receive_date&#39;; public $is_repeat = &#39;0&#39;; public $repetition_frequency_unit = NULL; public $repetition_frequency_interval = NULL; public $end_frequency_unit = NULL; public $end_frequency_interval = NULL; public $end_action = NULL; public $end_date = NULL; public $is_active = &#39;1&#39;; public $recipient_manual = &#39;&#39;; public $recipient_listing = &#39;&#39;; public $body_text = NULL; public $body_html = &#39;&lt;p&gt;sk cash&amp;nbsp;{contribution.financial_type_id:label}&lt;/p&gt;\n&#39;; public $sms_body_text = NULL; public $subject = &#39;yay&#39;; public $record_activity = &#39;1&#39;; public $mapping_id = &#39;saved_search&#39;; public $group_id = NULL; public $msg_template_id = NULL; public $sms_template_id = NULL; public $absolute_date = NULL; public $from_name = NULL; public $from_email = NULL; public $mode = &#39;Email&#39;; public $sms_provider_id = NULL; public $used_for = NULL; public $filter_contact_language = NULL; public $communication_language = NULL; public $created_date = &#39;2023-08-30 05:08:48&#39;; public $modified_date = &#39;2023-08-30 05:08:48&#39;; public $effective_start_date = NULL; public $effective_end_date = NULL; protected $resultCopies = 0; protected $_options = []; public $_DB_DataObject_version = &#39;1.11.3&#39;; public $__table = &#39;civicrm_action_schedule&#39;; public $N = 1; public $_database_dsn = &#39;&#39;; public $_database_dsn_md5 = &#39;320a00603e17f988282908defa13fbd6&#39;; public $_database = &#39;dmastercivicrm&#39;; public $_query = FALSE; public $_DB_resultid = 64; public $_resultFields = FALSE; public $_link_loaded = FALSE; public $_join = &#39;&#39;; public $_lastError = FALSE }</span>, <span>$phase = </span><span>&#39;rel-first&#39;</span>, <span>$defaultParams = </span><span>[&#39;casActionScheduleId&#39; =&gt; &#39;3&#39;, &#39;casMappingId&#39; =&gt; &#39;saved_search&#39;, &#39;casMappingEntity&#39; =&gt; &#39;civicrm_contribution&#39;, &#39;casNow&#39; =&gt; &#39;20230830052037&#39;]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/ActionSchedule/RecipientBuilder.php' bgcolor='#eeeeec'>.../RecipientBuilder.php<b>:</b>283</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>20</td><td bgcolor='#eeeeec' align='center'>31.6987</td><td bgcolor='#eeeeec' align='right'>5772712</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->getSql(  )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/ext/search_kit/Civi/Search/ActionMapping.php' bgcolor='#eeeeec'>.../ActionMapping.php<b>:</b>164</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>21</td><td bgcolor='#eeeeec' align='center'>31.7640</td><td bgcolor='#eeeeec' align='right'>6653912</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->buildWhereClause(  )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4Query.php' bgcolor='#eeeeec'>.../Api4Query.php<b>:</b>79</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>22</td><td bgcolor='#eeeeec' align='center'>31.7646</td><td bgcolor='#eeeeec' align='right'>6654608</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->treeWalkClauses( <span>$clause = </span><span>[0 =&gt; &#39;is_deceased&#39;, 1 =&gt; &#39;=&#39;, 2 =&gt; FALSE]</span>, <span>$type = </span><span>&#39;WHERE&#39;</span>, <span>$depth = </span>??? )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4SelectQuery.php' bgcolor='#eeeeec'>.../Api4SelectQuery.php<b>:</b>225</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>23</td><td bgcolor='#eeeeec' align='center'>31.7646</td><td bgcolor='#eeeeec' align='right'>6654608</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->composeClause( <span>$clause = </span><span>[0 =&gt; &#39;is_deceased&#39;, 1 =&gt; &#39;=&#39;, 2 =&gt; FALSE]</span>, <span>$type = </span><span>&#39;WHERE&#39;</span>, <span>$depth = </span><span>0</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4Query.php' bgcolor='#eeeeec'>.../Api4Query.php<b>:</b>213</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>24</td><td bgcolor='#eeeeec' align='center'>31.7647</td><td bgcolor='#eeeeec' align='right'>6654608</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->getExpression( <span>$expr = </span><span>&#39;is_deceased&#39;</span>, <span>$allowedTypes = </span><span>[0 =&gt; &#39;SqlField&#39;, 1 =&gt; &#39;SqlFunction&#39;, 2 =&gt; &#39;SqlEquation&#39;]</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4Query.php' bgcolor='#eeeeec'>.../Api4Query.php<b>:</b>243</td></tr>
<tr><td bgcolor='#eeeeec' align='center'>25</td><td bgcolor='#eeeeec' align='center'>31.7648</td><td bgcolor='#eeeeec' align='right'>6655112</td><td bgcolor='#eeeeec'>Civi\Api4\Query\Api4SelectQuery->getField( <span>$expr = </span><span>&#39;is_deceased&#39;</span>, <span>$strict = </span><span>TRUE</span> )</td><td title='/srv/civi-sites/dmaster/web/sites/all/modules/civicrm/Civi/Api4/Query/Api4Query.php' bgcolor='#eeeeec'>.../Api4Query.php<b>:</b>166</td></tr>

**Comment by eileenmcnaughton** [comment]
@colemanw perhaps we should also require `view all contacts` since we fetch the contact with `    $this->savedSearch['api_params']['checkPermissions'] = FALSE;` & that means no ACLs are used

I think without that the Greens / multisites could get in trouble on this. I'd rather be too strict at first & loosen later. There is also a 'view all custom data' I think that the greens do not give all admin users from memory (@seamuslee001 ?)

**Comment by seamuslee001** [comment]
@eileenmcnaughton I think that might be right but I would have thought that the is_deleted error should have been swallowed somewhere along the way

**Comment by eileenmcnaughton** [comment]
@seamuslee001 yeah the is_deleted is kinda separate - in that I looked in the code for that reason & then saw the checkPermission FALSE & wondered about the implications of one person configuring & assuming that what they see (their permissions) would be in play but instead someone elses is

**Comment by colemanw** [comment]
@eileenmcnaughton @JoeMurray I think before going further we should consider the pros and cons of restricting this new permission to only SearchKit communications. The unpermissioned query issue you've identified is endemic to the existing scheduled communications system which this new feature inherits. So for example if you create a (non-SearchKit) scheduled communication for `Contacts` it will bypass your personal ACLs and go out to *all* contacts meeting your criteria.
I understand the argument for a new feature hidden behind a new permission, but wonder if the extra work involved justifies it. It would be less dev time to require the same permissions for all scheduled communications, and it would plug the holes that are already there in addition to the new ones.

**Comment by eileenmcnaughton** [comment]
@seamuslee001 @johntwyman @andrew-cormick-dockery can you weigh in at all?

The issue is

1) we have a new feature that makes it possible to specify the critieria for a scheduled reminder in a search kit (ie gets away from the existing, weird, hard-coded field choices)
2) we have identified that at the point of sending permissions are not applied - ie a person who has ACL restrictions could have inadventently configured a scheduled reminder that will go to people they do not have permissions to see
3)@colemanw has pointed out this is the status quo
4) we have already decided to require a new permission for search-kit scheduled communications with the thinking we would require it for existing scheduled communication configuration later (giving a bit more testing time before most people start to see the new feature).

I had suggested that users need 'view all contacts' to configure scheduled reminders because any ACLS that they have are overridden when calculating them (new style or old). If we did that I think we would to both legacy and new style

I think @colemanw is suggesting that rather than require that we accept that anyone with 'Schedule Communications' permission can configure them to go to people they don't otherwise have access to - and require this permission for all scheduled reminders. This would probably look like a pre upgrade message saying that people will need to review their permissions and add this new permission to anyone who they want to be able to schedule reminders with some warnings about how it interacts with ACLS. This would probably be pretty clear on the upgrade screen but not that visible to people who might be assigning permisssions in 6 months time

In general people having administer civicrm and not view all contacts is a less common config - but one which is used in multisite & it seems like ideally we would tighten the permissions in this scenario with 1 or both of the above.

Note the current 'permission rollout' also has the effect of 'hiding' the new feature in the initial release - which I kinda liked as it would mean those who want to play can but most site admins would not be aware of it.

**Comment by andrew-cormick-dockery** [comment]
For us, scheduled reminders are used for two functions: membership and event reminders. We use multisite to constrain people's viewing contacts to Australia's states. But since membership is also state-based, and of course events are generally even more local than that, if scheduled reminders are completely unconstrained in that manner, that doesn't really bother us. I guess the fact that I'm aware of this now might help us avoid problems in the future.

If scheduled reminders are going to be expanded to use Search Kit, from our perspective even if it's not ACL constrained it's not going to bother us too much. This is because we've restricted the use of Search Kit to administrators due to the serious performance issues which could come about through a less restrictive regime.

But it sounds like you're intending to restrict that function to "view all contacts" in any case, which would be ideal. Only administrators have that function on our system.

(For what it's worth, if there's any part of scheduled reminders I'd like to see worked on, it's cleaning up the Manage Schedule Reminder screen! Oh, and if you're working on multiple reminders at a time in multiple tabs, the last one to save doesn't overwrite all the ones you have open!)

**Comment by colemanw** [comment]
> (For what it's worth, if there's any part of scheduled reminders I'd like to see worked on, it's cleaning up the Manage Schedule Reminder screen! Oh, and if you're working on multiple reminders at a time in multiple tabs, the last one to save doesn't overwrite all the ones you have open!)

For what it's worth, those issues have all recently been fixed!

**Comment by colemanw** [comment]
> As far as I understand, one of the purposes of this issue is to be able to send the results of a SearchKit by email using Scheduled Reminders like CiviReport does.

@francescbassas sorry no this PR does not do that. This PR allows you to use SearchKit to pick the recipients of the scheduled communication. It does not send search kit reports by email.

**Comment by eileenmcnaughton** [comment]
OK - moving the new feature stuff to an extension means we can merge it knowing only people who choose to play with it will have the feature - allowing us a bit more time to resolve the permissions question. All the core stuff has good test cover

**Comment by colemanw** [comment]
@eileenmcnaughton I pushed up a bit of error checking to ensure it would delete the reminder when deleting the saved-search, and to make sure it doesn't crash if you disable the extension.

**Pull request merged**